libs/constrained-generators/src/Constrained/Univ.hs

```diff
@@ -0,0 +1,318 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Constrained.Univ where
+
+import Constrained.List
+import Control.Monad.Identity
+import Data.Functor.Const
+import Data.Kind
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Set (Set)
+import Data.Set qualified as Set
+
+class FunctionLike fn where
+  -- | The semantics of a function is given by `sem`
+  sem :: fn as b -> FunTy as b
+
+class Member (fn :: [Type] -> Type -> Type) fnU where
+  -- | Inject a specific function symbol into a function
+  -- universe.
+  injectFn :: fn as b -> fnU as b
+
+------------------------------------------------------------------------
+-- Fixpoints
+------------------------------------------------------------------------
+
+{-
+On the most basic level what we are after is a first order representation of
+typed functions. Furthermore, we need the representation to be compositional so
+that you can plug-in your own functions into the universe.
+
+That might sound like it's no problem, but the issue is with higher-order
+functions. Consider:
+
+data ListFn where
+  MapFn :: _ '[a] b -> ListFn '[[a]] [b]
+
+What should go in place of `_`? It turns out that you need to specify the global
+function universe and make sure the global universe is the same everywhere.
```

Comment:
.. and make sure the same global universe applys to every element of the List as well (hence the '[[a]])

Reply:
The list `'[[a]]` here is just saying that the function takes `[a]` as argument. At this point the arguments have nothing to do with the universe at all.